crc: add popular crc8  and crc32 variants

Igor-Misic opened the pull request:
I've changed our crc32 implementations to crc32_base so it can calculate any crc32. I've also added a function that can calculate the lookup table and then use crc32_base to add support for some common crc32 variants. 
This page http://www.sunshine2k.de/coding/javascript/crc/crc_js.html is used as a source for popular variants. 

The same is done for crc8. Here is a list of supported variants:

##### CRC8
- CRC8       // Basic
- CRC8 8H2F
- CRC8 CDMA2000
- CRC8 DARC
- CRC8 DVB S2
- CRC8 EBU
- CRC8 ICODE
- CRC8 ITU
- CRC8 MAXIM
- CRC8 ROHC
- CRC8 SAE J18850
- CRC8 SAE J18850 ZERO
- CRC8 WCDMA

##### CRC32
- CRC32      // Basic
- CRC32 BZIP2
- CRC32 C
- CRC32 D
- CRC32 JAMCRC
- CRC32 MPEG2
- CRC32 POSIX
- CRC32 Q
- CRC32 XFER

This way we can quickly add new variants if needed and users could use the existing variants with no knowledge about how to implement the new one or the advanced users could use  "base" to implement any other variant. 

I've planning to do the same for CRC16.

jciberlin reviewed (commented):
First of all, great work! Furthermore, I have some general comments. Maybe it will be better to avoid separate header/source files for every crc variant. All variants are calling/using the same function. So, we can have one header file where we will define crc_tables for each variant, and have only one crc function which will be called with different parameters. In that header file we can define initial xor value, final xor value, etc. as well. Maybe we can have ifdef for each variant and user can use different variant by defining particular variant in makefile. Also, for function names it will be better if we follow something like this: Crc32_functionName, and for file names: crc32_mpeg2.c.

jciberlin commented on Src/crc/crc32/crc32-jamcrc.c:
```diff
@@ -0,0 +1,99 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc32/crc32-mpeg2.c:
```diff
@@ -0,0 +1,99 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc32/crc32-q.c:
```diff
@@ -0,0 +1,99 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc32/crc32-xfer.c:
```diff
@@ -0,0 +1,99 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8-cdma2000.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8-darc.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8-dvb_s2.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8-ebu.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8-icode.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8-maxim.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8-rohc.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8-sae_j1850_zero.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8-wcdma.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

jciberlin commented on Src/crc/crc8/crc8.c:
```diff
@@ -0,0 +1,83 @@
+#if (FINAL_XOR_VALUE == 0x0U)
+               false
+#else
+               final_xor
```
Undefined variable if FINAL_XOR_VALUE is different than 0x0U.

Igor-Misic commented:
I updated the function to follow our current code style. Thx!

>and for file names: crc32_mpeg2.c.

Yeah, my intention was to somehow emphasize variants from crcX_base, but it seems I have failed :D
I changed it, let me know if it looks more intuitive now.

> Maybe it will be better to avoid separate header/source files for every crc variant. 

I was thinking about it at the beginning but in the end, I decided to do it in separate files. 
There is a couple of reason from my side why I decide to do it this way:
1) I like more small files
2) we would like to be able to enable all variants (for users and unit tests)
3) Beceose of 2) the only part that will be common for all files is `CrcX_base() ` function call
4) Less #defines MACROS for users to think about with means less documentation. Now it is straightforward, you include the source you use and it is visible from Makefile. 
5) We don't need to force tastes on each other in the project. A bit of variety is not always bad unless it doesn't follow MISRA xD

Anyhow, I don't discourage the idea that we can use #ifdefs in the future.

Igor-Misic committed:
crc: add popular crc8  and crc32 variants
